### PR TITLE
[sonos] Avoid concurrent execution of the polling job

### PR DESCRIPTION
**Pull request by lolodomo**
Optimize the UPnP calls in the polling job

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
Jar for testing:
[org.openhab.binding.sonos-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4516747/org.openhab.binding.sonos-2.5.5-SNAPSHOT.zip)

File: bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java [hunk 1]
```diff
@@ -144,38 +144,42 @@
 
     private final Object stateLock = new Object();
 
+    private final Object jobLock = new Object();
+
     private final SonosStateDescriptionOptionProvider stateDescriptionProvider;
 
     private final Runnable pollingRunnable = () -> {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:151]: While you are changing code here anyway: Can you convert the runnable to a `private void poll()`?

Comment (lolodomo) [reply bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java]: Done

File: bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java [hunk 2]
```diff
@@ -100,96 +98,63 @@
...
     public ZonePlayerHandler(ThingRegistry thingRegistry, Thing thing, UpnpIOService upnpIOService, String opmlUrl,
             SonosStateDescriptionOptionProvider stateDescriptionProvider) {
         super(thing);
         this.localThingRegistry = thingRegistry;
         this.opmlUrl = opmlUrl;
         logger.debug("Creating a ZonePlayerHandler for thing '{}'", getThing().getUID());
-        if (upnpIOService != null) {
-            this.service = upnpIOService;
-        }
+        this.service = upnpIOService;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:157]: I agree that the null-check like it was before does nothing,  but I wonder if removing is the correct way. Are you sure that every access to `service` is null-safe? Otherwise it would be probably better to throw an `IllegalArgumentException`. 

The correct way would obviously be to add proper null-annotations, but that is out-of-scope of this PR.

File: bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java [hunk 3]
```diff
@@ -230,6 +195,41 @@ public void initialize() {
         }
     }
 
+    private void poll() {
+        synchronized (jobLock) {
+            try {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:200]: You should add some logic here to check whether or not this handler was disposed while waiting to acquire the jobLock.
```suggestion
        synchronized (jobLock) {
            if (pollingJob == null)  {
                return;
            }
            try {
```

File: bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:157]: The Sonos handler factory bundle will not start without this bundle. It is set as a reference:
https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/SonosHandlerFactory.java#L74
So it cannot be null.
That is exactly the same thing as for thingRegistry or stateDescriptionProvider.

File: bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java (hunk 3 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:200]: No, because this method is finally not called only by the polling job.
It is what I just discovered and why I added this synchronization.

Comment (lolodomo) [reply bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:200]: This method is not called only by the polling job.
I see what you mean, we should avoid to run it just after the thing handler was disposed.

Comment (lolodomo) [reply bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java]: Done

**Comment by lolodomo** [comment]
New jar for testing:
[org.openhab.binding.sonos-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4527380/org.openhab.binding.sonos-2.5.5-SNAPSHOT.zip)

**Comment by lolodomo** [comment]
New jar for testing:
[org.openhab.binding.sonos-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4528982/org.openhab.binding.sonos-2.5.5-SNAPSHOT.zip)

**Comment by J-N-K** [comment]
@cpmeister Can you check if everything is addressed?

File: bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java [hunk 4]
```diff
@@ -935,46 +939,31 @@ protected void updateRunningAlarmProperties() {
...
-            update = true;
+        if (StringUtils.isNotEmpty(this.stateMap.get("IPAddress"))) {
             properties.put(IP_ADDRESS, this.stateMap.get("IPAddress"));
         }
-        if (update) {
-            updateProperties(properties);
-        }
+        updateProperties(properties);
+
+        return !result.isEmpty();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:966]: Why did you change the way this was implemented? It looks like it was intended to update properties if they haven't changed. Now you are updating the properties every time.

Comment (lolodomo) [reply bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:966]: I changed it because this was useless, the method `updateProperties` from the core framework is already checking if there is one change before updating the properties.

File: bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java [hunk 5]
```diff
@@ -100,105 +98,74 @@
...
     @Override
     public void dispose() {
         logger.debug("Handler disposed for thing {}", getThing().getUID());
 
-        if (pollingJob != null && !pollingJob.isCancelled()) {
-            pollingJob.cancel(true);
+        if (pollingJob != null) {
+            if (!pollingJob.isCancelled()) {
+                pollingJob.cancel(true);
+            }
```
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java:168]: Done

**Comment by cpmeister** [comment]
Manually checked sign-off.
